src/main/telemetry/ibus_shared.c

```diff
@@ -57,120 +79,460 @@ typedef enum {
+    IBUS_SENSOR_TYPE_ARMED            = 0x15, //2 bytes
+    IBUS_SENSOR_TYPE_FLIGHT_MODE      = 0x16, //2 bytes
+    IBUS_SENSOR_TYPE_PRES             = 0x41, // Pressure
+    IBUS_SENSOR_TYPE_ODO1             = 0x7c, // Odometer1
+    IBUS_SENSOR_TYPE_ODO2             = 0x7d, // Odometer2
+    IBUS_SENSOR_TYPE_SPE              = 0x7e, // Speed 2bytes km/h
+
+    IBUS_SENSOR_TYPE_GPS_LAT          = 0x80, //4bytes signed WGS84 in degrees * 1E7
+    IBUS_SENSOR_TYPE_GPS_LON          = 0x81, //4bytes signed WGS84 in degrees * 1E7
+    IBUS_SENSOR_TYPE_GPS_ALT          = 0x82, //4bytes signed!!! GPS alt m*100
+    IBUS_SENSOR_TYPE_ALT              = 0x83, //4bytes signed!!! Alt m*100
+    IBUS_SENSOR_TYPE_ALT_MAX          = 0x84, //4bytes signed MaxAlt m*100
+
+    IBUS_SENSOR_TYPE_ALT_FLYSKY       = 0xf9, // Altitude	2 bytes signed in m
+#if defined(TELEMETRY_IBUS_EXTENDED)
+    IBUS_SENSOR_TYPE_GPS_FULL         = 0xfd,
+    IBUS_SENSOR_TYPE_VOLT_FULL        = 0xf0,
+    IBUS_SENSOR_TYPE_ACC_FULL         = 0xef,
+#endif //defined(TELEMETRY_IBUS_EXTENDED)
+    IBUS_SENSOR_TYPE_UNKNOWN          = 0xff
 } ibusSensorType_e;
 
-/* Address lookup relative to the sensor base address which is the lowest address seen by the FC
-   The actual lowest value is likely to change when sensors are daisy chained */
-static const uint8_t sensorAddressTypeLookup[] = {
+typedef union ibus_telemetry{
+    uint16_t uint16;
+    uint32_t uint32;
+    int16_t int16;
+    int32_t int32;
+    uint8_t byte[4];
+} ibus_telemetry_t;
+
+#if defined(GPS)
+
+const uint8_t GPS_IDS[] = {
+    IBUS_SENSOR_TYPE_GPS_STATUS,
+    IBUS_SENSOR_TYPE_SPE,
+    IBUS_SENSOR_TYPE_GPS_LAT,
+    IBUS_SENSOR_TYPE_GPS_LON,
+    IBUS_SENSOR_TYPE_GPS_ALT,
+    IBUS_SENSOR_TYPE_GROUND_SPEED,
+    IBUS_SENSOR_TYPE_ODO1,
+    IBUS_SENSOR_TYPE_ODO2,
+    IBUS_SENSOR_TYPE_GPS_DIST,
+    IBUS_SENSOR_TYPE_COG,
+};
+#endif
+
+#if defined(TELEMETRY_IBUS_EXTENDED)
+
+const uint8_t FULL_GPS_IDS[] = {
+    IBUS_SENSOR_TYPE_GPS_STATUS,
+    IBUS_SENSOR_TYPE_GPS_LAT,
+    IBUS_SENSOR_TYPE_GPS_LON,
+    IBUS_SENSOR_TYPE_GPS_ALT,
+};
+
+const uint8_t FULL_VOLT_IDS[] = {
     IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE,
-    IBUS_SENSOR_TYPE_TEMPERATURE,
-    IBUS_SENSOR_TYPE_RPM
+    IBUS_SENSOR_TYPE_CELL,
+    IBUS_SENSOR_TYPE_BAT_CURR,
+    IBUS_SENSOR_TYPE_FUEL,
+    IBUS_SENSOR_TYPE_RPM,
 };
 
-static serialPort_t *ibusSerialPort = NULL;
+const uint8_t FULL_ACC_IDS[] = {
+    IBUS_SENSOR_TYPE_ACC_X,
+    IBUS_SENSOR_TYPE_ACC_Y,
+    IBUS_SENSOR_TYPE_ACC_Z,
+    IBUS_SENSOR_TYPE_ROLL,
+    IBUS_SENSOR_TYPE_PITCH,
+    IBUS_SENSOR_TYPE_YAW,
+};
 
-#define INVALID_IBUS_ADDRESS 0
+#endif //defined(TELEMETRY_IBUS_EXTENDED)
+
+static serialPort_t *ibusSerialPort = NULL;
 static ibusAddress_t ibusBaseAddress = INVALID_IBUS_ADDRESS;
+static uint8_t sendBuffer[IBUS_BUFFSIZE];
+
+
+static void setValue(uint8_t* bufferPtr, uint8_t sensorType, uint8_t length);
```

Comment:
Try to avoid forward declarations by reordering function definitions instead.

Reply:
@mikeller can not be done because setCombinedFrame uses setValue and setValue  uses setCombinedFrame.